### ISSUE TITLE
Ignore video progress while using PseudoTV

### ISSUE THREAD

**Issue by yattaah** [Steveb1968/script.pseudotv#7]
First of all thanks for your fork - this is exactly what a lot of people needed,
I always had the problem that PseudoTV changes the watch status of the played videos (which makes sense in general). For example after switching to another channel in de middle of the movie, the movie is set to in progress in the library (or even finished when the movie ended on the channel). 

Is there a way to prevent this? I would really prefer to seperate the library watch status from PTV.

**Comment by Steveb1968** [comment]
I'm sure there is a way using the json api, but I'm not sold on the benefits of this.
If is watched/in progress, kodi's db should reflect that.
There is a Advanced channel rule "Only Watched" that could be useful in some cases.
I'll have a think about this a maybe add as a option.

**Comment by yattaah** [comment]
My current workaround is to create channels with an "Only Watched" rule (to prevent spoilers) and to prevent Kodi logging the status (via advancedconfig - see: http://kodi.wiki/view/HOW-TO:Modify_automatic_watch_and_resume_points). My "real" watched status is synced via trakt. The disadvantage is of course that I have to sync via trakt every time I watched somethinng via the library.

This would still be a very nice option to have - when someone watches a channel for 5 minutes the video (for example an episode) is set to in progress. But when you already watched this ep (or you still have to watch it) the status in the db is kind of irritating. They also appear on the homescreen which doesn't help either (because they are not "in progress" by user's choice).

**Comment by Steveb1968** [comment]
Yep, I get where your coming from. I will put this on the short list as a settings option.

**Comment by Remco32** [comment]
> but I'm not sold on the benefits of this.

My content is in a different location. I stream my content using Plex. There's a Kodi add-on which allows you to add your Plex content as regular video files to your library.

I haven't set up PseudoTV here. But if I were to do that, it would confuse Plex: it would think I'm halfway through a bunch of episodes. Simply because I would be flicking channels in PTV.

**Comment by Steveb1968** [comment]
I will have a look at this when I can get some time.
In the meantime have a look at this, it could help.
[http://kodi.wiki/view/HOW-TO:Modify_automatic_watch_and_resume_points](http://kodi.wiki/view/HOW-TO:Modify_automatic_watch_and_resume_points)

**Comment by yattaah** [comment]
Any updates on this option? I'm currently in the progress of creating a new setup und would love to use PseudoTV without "breaking" (or better unexpectedly changing) my progress stats in my library / the widgets (e.g. in progress) on my home screen.

**Comment by Steveb1968** [comment]
Hi yattaah, no sorry. I was hoping to get to it on the holiday break but it's all over already. (WTF) :)
It's still on the list. It's not a easy feature to add if you want to get it right. ;)
Whats your main concern. Watched status? Resume? Or both?  :)

Cheers

**Comment by yattaah** [comment]
No Pressure! =)

Yeah, I can imagine that it's hard to implement (if even possible?)
The watch status and resume points should just stay untouched but they should still wotk while using Kodi just normally (starting a movie in the library, etc.). 

Sadly I have absolutly no idea how this could me achieved smoothly...

And thanks again for all your work, you're a really nice guy! :D

**Comment by fnord12** [comment]
I took a crack at this.  My approach was to write the initial Playcount and Resume information to the m3u cache files, and then when closing PseudoTV use JSON to write those values back to each video in the caches.  And... it works!   But unfortunately it takes a really long time - over a minute for a mere 200 videos across 4 channels.  I'm a terrible newbie programmer and I'm sure my code could be made more efficient, but it's clear that it's the JSON calls that are taking the time (if I comment out just the JSON lines and leave in all my terrible code, everything else happens near instantaneously).

I thought about trying to run these updates as a background task that could continue even after PSTV closes, but I don't really know how to do that and I'm now doubting that this is the right approach.

Just thought I would report my attempt.  I can share my code if anyone is interested.

**Comment by Steveb1968** [comment]
@fnord12. Hi mate and respect for having a crack. Your definitely on the right track, JSON is the way to go.
Grab the data you need (play count, resume point, dbid) and store that to a temp file or memory. Then on exit if the value has changed set it back to the original value.
I thought about using the cached m3u files too but they could hang around for a while if you don't refresh your channel list on startup.
I think it can be done but as you've found out this comes at a cost and that is time. :)

If you want to share what you've done, go for it. Don't worry about how good your code is. Your having a go and that's all that matters. :)

**Comment by fnord12** [comment]
Thanks Steve especially for your encouragement.  You can (hopefully) see my changes here:
https://github.com/fnord12/script.pseudotv/pull/2/files

I did it mostly along the lines you said, but I'm not checking if the value had changed before writing back.  I thought adding another round of JSON queries to check the current values would just increase the time further.  But maybe I should try that.  I can also see the value of writing to separate files instead of the m3us since it would reduce the amount of parsing of the file that has to happen on exit (but parsing the files doesn't seem to be where the time is coming from).

I know you're busy and so no hurry on any feedback.

**Comment by Steveb1968** [comment]
@fnord12 Nice work mate, I can see you put a lot of time into that.. and it works. :)
Shutdown time is a problem but hey if it works for you. ;)
Watchout though, coding problems like this can become addictive.. :) That's why I have a coding ban on me at the mo.. :(   :)

**Comment by Steveb1968** [comment]
Side note: Json library queries don't take too long we make a bunch at startup along with writing all the m3u files and ref playlists. The parsing & the json.set stuff is where time is coming from.
Resume point is also a big hit, I wasn't too keen to add this personally.
If watched count > 0 "forget about it" :) else set it to its original value.
That was my inital thoughts on the subject but hey, go for gold mate! :)

**Comment by fnord12** [comment]
Thanks again Steve.  Yeah, unfortunately the shutdown time makes it unusable.  I'll give it another try with an extra round of JSON queries to avoid so many writes, and I'll see if I can find a way to reduce writing back all the resume points (the goal is to not trigger the movies sampled in PSTV showing up on "continue watching" type lists, so I think any progress on the resume might affect that).  And the parsing didn't _seem_ to be taking any noticeable time when I commented out the JSON writes, but I can try making that more efficient too (maybe reading the m3u files backwards at the end instead of having to re-parse out the titles, plots, etc,, or maybe writing to a separate file).

I really appreciate your feedback; it's given me a lot to go on.

**Comment by Steveb1968** [comment]
No probs mate, happy to help! I think the resume point just needs a check for change to avoid the json.set. I think that would speed things up. If you want I could create a test branch where I could pull in your updates and work from there.

**Comment by Steveb1968** [comment]
Created a Test-branch based on master. So if you want you could fork that, merge your updates and send a pull request.
If not, that's cool!!  I understand. keep it local. ;)

**Comment by fnord12** [comment]
I'm very happy to have you continue from what I've attempted (if it doesn't interfere with your self imposed code ban!).  I've created a pull request with my current work merged with your test branch.  If I've misinterpreted and you don't want it until I attempt what you've suggested, just ignore/reject it.  Also apologies if I've screwed up the PR somehow; it's my first time submitting to someone else's code.

**Comment by Steveb1968** [comment]
Hi mate, unfortunately it looks like your fork is a commit behind master. If you could rebase your fork and merge your changes that would be good. Otherwise it will revert the latest changes in master.
No rush mate, I haven't much time on my hands at the moment. ;)

**Comment by fnord12** [comment]
Weird, I (thought I) forked your test-branch, uploaded my files, and resolved the conflicts and then did a pull request into the test-branch.  Looks like I need to find a Github tutorial.  ;-)   I'll try again but probably not until next week.

**Comment by Steveb1968** [comment]
Yea mate, all good. Like I said no rush. I think the problem your having is the method your using to update the fork. Uploading out of date files is not the way to go. Try to work with your fork directly and commit to that, then when your ready send a pull.
GitHub can take some time to get your head around.. :)

**Comment by rafaelvieiras** [comment]
We have some news about this feature?

:smile:

I use emby for a library manager! So, I created a user "tvwatcher" for not change my user progress while use PTV!

**Comment by fnord12** [comment]
At this point i've got fully working code that resets the statuses in a reasonably short period of time.  I've been using it for about a month now and haven't run into issues.  I've been gunshy about loading it back up as a pull request after my last fail, but i'll try to do it tomorrow.

**Comment by Steveb1968** [comment]
Cool mate sounds good. With GitHub if your fork gets behind master you can rebase your fork so it is equal with master while stashing the changes you made, once rebased you then push your stashed changes back to your fork. I hope that makes sense, it can get a bit confusing. :)
Give it a go anyway I'm sure we'll figure it out.

**Comment by rafaelvieiras** [comment]
Good @fnord12 , if you need help, talk to me! http://t.me/rafaelvieiras

@Steveb1968, I think is a good idea if you create a telegram group or a https://gitter.im/ channel for this repository! What do you think?

**Comment by Steveb1968** [comment]
I have a look, I must admit I have not heard of gitter.im before.
Sounds like a good idea though. ;)

**Comment by fnord12** [comment]
Gave it another try.

**Comment by rafaelvieiras** [comment]
I Put a comment on your PR @fnord12 !

**Comment by fnord12** [comment]
Sorry guys.  Definitely struggling a bit with GitHub, but I think I've cleared it up with the new Pull Request.  Thanks for your patience.

**Comment by Steveb1968** [comment]
Thanks for that. Merged to test branch and ready for a test run. 😁

**Comment by Steveb1968** [comment]
Nice work mate! Works well and the shutdown time is reasonable. I tested with a virgin portable kodi 18.5 install with default PseudoTV settings.
I only have a small test library on my dev machine so with default it creates 17 channels. I bumped the playlist limit to 100 and flicked through all channels.
The extra time on exit is totally reasonable for a setting that will improve you library experiance. I'm sure you could bring that time down even further too. 

I've got a few suggestions to improve it, that's if you still want to work on it?
The main one is writing to the m3u file. With a 100 movie list and auto settings that m3u could stick around for close to a week without being refreshed. Something to think about. :)

**Comment by Steveb1968** [comment]
I think I'll push this to master to hopefully get a bigger test group. :)

**Comment by yattaah** [comment]
Great work guys! And especially thank you fnord12, I really appreciate it! =)

I switched my media center setup (not using Kodi anymore) but I will test drive this and if it works well (and I'm sure it does) I will switch my whole setup back to Kodi again. :D

**Comment by Steveb1968** [comment]
@fnord12 This is know in master, work from there if your still interested.

**Comment by fnord12** [comment]
Thanks for the feedback Steve, and looking forward to more from yattaah and others who try it.  I'm sure there have to be bugs in there.  Still wrapping my head around why it's a problem to write to the m3us; I guess the alternative is writing to a temp file just for that session, but I'm not sure if I'm capable of tackling that.

**Comment by Steveb1968** [comment]
Personally I would have just appended the info to a list and not written to file. see rules.py for a example of that. Then the data collected is always current on each run and discarded at shutdown. The m3u files are used to store more permanent data (duration, title, plot etc.) and is used in the info and epg windows. Depending on settings the m3u is not rewritten on startup and is used to speed up channel creation it will be refreshed when the channel is running out of content to play.
All good though, it works.

**Comment by yattaah** [comment]
Okay, I did a testrun but for whatever reason the status doesn't reset on my end (I get the notification and I see the typical flickering that you get when kodi is editing a lot of entries). I'm sure the problem is on my end.

It is a completely fresh setup. I added my libraries (Zeroconfig SMB), updated watch status via trakt, turned on reset watch status in PTV settings and then I startet Pseudo TV. Switched through some channels and then exit again. All videos that started are in progress on my homescreen/library. 

I wasn't able to find the problem in the debug log - I attached it, maybe it helps.

[kodi.log](https://github.com/Steveb1968/script.pseudotv/files/3882407/kodi.log)

**Comment by fnord12** [comment]
Hi yattah.  Thanks for a great explanation of what you did and for the log.

In the log, I see what looks like 3 successful resets, two for movies with IDs 114 and 97 and one for TV episode "Subtle Beast" (1x02).  I also see it going through the rest of the channel playlists with no actions.  Are those the episodes that show up as "in progress"?  I'm trying to determine if the problem is that those 3 failed to properly reset, or if it should have reset other videos.

**Comment by rafaelvieiras** [comment]
It's not working for me. :cry: 
I'm using Emby Server to media delivery, and this technic not work in emby api, only local kodi database!

**Comment by yattaah** [comment]
Alright, I run a few more tests and listed the videos that were running and compared them to the playstate after exiting. I think a see the pattern that already watched items get a reset, but unwatched items get flagged as in progress. The three items you list got the reset btw, so this works fine =)

I would prefer to reset everything (watched and unwatched), but maybe some users could prefer to flag items as watched after watching unplayed items in PTV. (like on TV, zapping around, stay on a channel because you're interested and after that you count it as watched).

Would it make sense to default on "reset everything" and to add an optional setting as in "mark unwatched videos as watched after finishing"?

**Comment by fnord12** [comment]
yattaah, the code shouldn't be excluding resets basd on watched status, exactly, but it did have a check that was only supposed to reset videos where the current (JSON) playcount and resume weren't at 0.  I've removed that check and made one other change and submitted a pull request for Steve.  Hopefully you can give it another try if he accepts it.

**Comment by fnord12** [comment]
rafael, I don't use Emby so I don't know if what I've done even can work for it, and I'm not able to troubleshoot directly.  Two things to try:

1. Can you look at your cached m3u files and tell me if the resume data is getting written to the end of each entry?  The files are in your userdata folder under addon_data\script.pseudotv\cache.  If the data is getting written, you'll see something like this after each plot description:
//1//0.0//2019-10-16 20:41:17//42077

2. Can you install this simple add on from my repository: https://github.com/fnord12/script.resetwatched

This will add a Reset Watched option to your context menu.  You can test if it works by playing a movie, jumping to some point in the middle, and then quitting.  Then restart the movie and confirm it's starting from a resume point.  If so, try clicking Reset Watched and then starting the movie again.  It should start from the beginning without offering to resume.

I use the same JSON call when resetting in my PseudoTV code, so this will help confirm if it works with Emby.

**Comment by psicotron3** [comment]
This is a really good and long waited enhancement, but I saw a kind of regression as I discovered some videos aren't being added to channels anymore. 
I'm on Windows 10, x64 and Leia 18.5. 
Everything was working fine on the last commit before ResetWatched.
Trying to find out what is happening, it seems the issue isn't related do filename or mediainfo metadata. 
I did a fresh run using both the last working version and the last one, adding only two movies into my library, with PseudoTV default options and setting "force channel reset" before each test run.
The result is that with the old version the 2 movies got added to a channel and the addon worked fine but the new version just exited with the error saying it wasn't able to populate the channel.
I also found out that, with the last version, none of those movies got added to the m3u file generated on cache folder (its completely empty).
I couldn't find anything any clue within the logs, but I'm attaching both of then anyway. 
If you need some other info, please let me know and thanks for the great plugin.

[kodi - not working.log](https://github.com/Steveb1968/script.pseudotv/files/4014873/kodi.-.not.working.log)
[kodi - working.log](https://github.com/Steveb1968/script.pseudotv/files/4014874/kodi.-.working.log)

**Comment by fnord12** [comment]
Sorry it's not working, psicotron3.  Looking at your logs, I see that in the non-working version, a little before the "fatal error", it calls "clearPlaylistHistory" and then says "channel not valid, ignoring", and that's seemingly what's causing the playlist to be empty and the movies to not play.  This portion of the code wasn't changed with my Reset Watched functionality.  That doesn't mean it isn't causing the problem indirectly, but it's going to be harder for me to troubleshoot.  I need to get familiar with that code.

I know that you said that the filename and mediainfo don't seem to be related, but can you please let me know the file extensions for the movies that aren't playing?

I also see that you're using a portable install, and when I wrote this, I was on 18.4 and haven't upgraded yet.  So there are differences there that I can investigate, but it will take a while.

**Comment by psicotron3** [comment]
I tried to "mimic" your build installing kodi 18.4 and using in regular "non portable mode". I also added some episodes from TV Shows: Mr. Robot (5 episodes), The Expanse (3 episodes), The Mandalorian (2 episodes) and Watchmen (5 episodes). Channels are played including only Mandalorian and Watchmen. I tried to add a channel for Mr. Robot and The Expanse but the m3u was created empty.

The issue with Movies is the same, and both movies I have with Adventure genre are not added to any channel (specifically the Adventure genre channel).

Checking the logo (attached), I can see that on line 634 you can see the JSON response to the config script and it indicates all the TV Shows in library, movies genres that I have added.

Line 3142 shows the JSON response with 5 episodes for the query to create channel based on smart playlist "Show_Mr. Robot_random.xsp" 

The following lines indicate the addon determining the episodes length and on line 3195 we see again the "clearPlaylistHistory" "channel not valid, ignoring" error as you stated before. So it confirms that you were right and those errors are related.

But, removing that check on ChannelList.py only stops the error from being displayed but changes nothing in the behavior. I think that the error comes from the detection of the empty m3u files and maybe some of the additions to the m3u creation have triggered the issue. If possible i'll make more tests to try to more specific about what are the conditions to recreate the problem.

I need to say that the "empty" m3u files I mention are generated only with the line "#EXTM3U" (not entirely an empty file).

I thought it could be related to short number of episodes, but mr. robot has 5 episodes and don't generate channel and mandalorian has only 2 and the channels are generated.

I also thought it could be related to watched flags and resumes points but the tests were made on a clean library.

All files are mkv and I copied the working files with different names as episodes on other series and movies, and the result is inconsistent (sometimes work, others not but never work on movies).

For sure, I'll try to figure it out because this feature is really needed.

[kodi.log](https://github.com/Steveb1968/script.pseudotv/files/4021031/kodi.log)

**Comment by fnord12** [comment]
Thanks for the additional details, psicotron3; I appreciate your help.  Can you replace the attached ChannelList.py file? It will write additional information into the log.  Can you turn on debugging mode in Kodi and then run PseudoTV with just the two action movies that don't load?  Hopefully this will give us a better idea of what's going on.

[ChannelList.zip](https://github.com/Steveb1968/script.pseudotv/files/4024010/ChannelList.zip)

**Comment by psicotron3** [comment]
Thanks for the interest.

Here it is:
[kodi-debug.zip](https://github.com/Steveb1968/script.pseudotv/files/4024652/kodi-debug.zip)

**Comment by fnord12** [comment]
Ok, the code is throwing an exception, but the code is in a Try/Except block that just passes files that throw an exception without giving more detail.  In the new attached ChannelList.py, I've taken the relevant code out of the Try/Except block.  In doing so, I've found a few reasons why the exceptions might occur, and I've fixed them.  So this may actually fix the problem for you.  If not, it will show more details about the exception.  But it will also cause Kodi to lock up.  So please turn on debugging and run PsTV with this file, and then grab the log before forcing Kodi to close and re-opening it (because that will wipe the log).

(One thing that will definitely cause this to fail is if you have an Advanced Rule excluding unwatched episodes.  I don't think you're using Advanced Rules for this test, but just FYI.)

[ChannelList.zip](https://github.com/Steveb1968/script.pseudotv/files/4028509/ChannelList.zip)

**Comment by psicotron3** [comment]
It worked. Really fine by the way. Here are the logs.
[kodi20200106.zip](https://github.com/Steveb1968/script.pseudotv/files/4028842/kodi20200106.zip)

I also tested with the TV Shows and everything is working fine right now. 

Thank you. Great work.

**Comment by fnord12** [comment]
Great.  I need to put the code back into the Try/Except block and then integrate it into the repository.

I introduced a problem with my RegEx expression when parsing the JSON response for the ID field.  Short story is it wasn't grabbing movies/episodes where the Database ID field was only 1 digit.  My library has gone through enough iterations that I didn't have many such values, but that's probably what was happening with yours.  People with larger libraries might not even have noticed that these movies were being skipped, so I'm glad that you caught it.

There was also older code that would cause an exception for Episodes that didn't have a title, and I fixed that too.  But that's probably a rarity that wasn't affecting most people (for me it was a few obscure foreign language shows that didn't have good TVDB data).

**Comment by undecomposed** [comment]
This is a bit old but I'll add my +1 for this feature. I often just want to put the PseudoTV in background so I can have some noise, but I still want to be able to watch the shows and movies in the future.

**Comment by fnord12** [comment]
This feature has been delivered.  It's slow, but it works, in Steveb's branch, and i've recently made improvements to make it much faster in my branch: https://github.com/fnord12/script.pseudotv

I'm going to close this issue since the feature has been implemented.

**Comment by clkish1988** [comment]
Any chance or interest in making this setting on a per channel basis? I have my main channel that I like to watch set to play unwatched episodes in order. I like to have other channels but if a show on my main channel plays on another channel the play flag gets set and my main channel won't play it...or do the play counts only count for actively watched shows?
